### DISCUSSION TITLE
Netboot Debian on a 2006 Xserve with iPXE

### DISCUSSION THREAD

**Discussion by robtech21** [ipxe/ipxe#1112] [General]
Hi, I am in ownership of a late 2006 Xserve, and I stumbled across [this article from 2020](https://pixilic.com/debian-headless-2009-xserve) about booting Debian 10 (a bit outdated, but from the URLs, it should netboot Debian 12) on an Xserve 1,1. Following the directions up until I type `boot` into the console and hit enter, I get the following message:

```
PXE net0 START IPv4
PXE net0 using 192.168.1.4
PXE net0 could not install base code protocol: Error 0x7f4ae082 (https://ipxe.org/7f4ae082)
Could not boot: Error 0x7f4ae082 (https://ipxe.org/7f4ae082)
```
(I did compile iPXE with `DEBUG=efi_pxe` to get this info after the initial error)
This is a bit obscure, as I can't really gleam anything from the provided link, other than [where it originates from](https://github.com/ipxe/ipxe/blob/master/src/interface/efi/efi_pxe.c#L1637), but I was hoping someone could help me out? Thanks!

**Comment by mcb30** [comment]
Interesting!  The error code indicates that the call to `bs->InstallMultipleProtocolInterfaces()` is returning `EFI_INVALID_PARAMETER`.

There isn't any obvious reason why it should be considered to be invalid.  As a first cut, you could try commenting out each of the lines
```c
&efi_pxe_base_code_protocol_guid, &pxe->base,
```
and
```c
&efi_apple_net_boot_protocol_guid, &pxe->apple,
```
in turn, so that we at least know which protocol the firmware is complaining about.

**Comment by robtech21** [comment]
I'll try that and let you know!

**Comment by robtech21** [comment]
Commented out the following line:
`&efi_apple_net_boot_protocol_guid, &pxe->apple,`
And it worked! It dropped me into the Debian netboot installer. Thanks! Debian installer doesn't appear to be hanging, but after this point it's really not any issue with iPXE...

**Comment by mcb30** [comment]
That's great!  It would be good to debug further if you get the chance.  I don't have any Apple hardware to test with, so that piece of code was taken largely on trust.

Some possibilities for the failure cause:

- Apple firmware may be calling in to our protocol implementation and not liking the values that we return.  You could add some debug prints in `efi_apple_get_dhcp_response()` and `efi_apple_get_bsdp_response()` to check if we're being called.  (Use the `DBG()` macro and build with `DEBUG=efi_pxe`).

- Apple firmware may be installing its own instance of `EFI_APPLE_NET_BOOT_PROTOCOL` onto the handle as soon as we install `EFI_PXE_BASE_CODE_PROTOCOL`.  If so, then commenting out only the `&efi_pxe_base_code_protocol_guid, &pxe->base,` line should also give you a working boot.

- Apple firmware may complain if more than one instance of this protocol is installed.  This one is slightly tricker to debug, but you can use e.g.
  ```c
  static EFI_APPLE_NET_BOOT_PROTOCOL *existing;
  EFI_REQUEST_PROTOCOL ( EFI_APPLE_NET_BOOT_PROTOCOL, &existing );
  ```
  to cause iPXE to search for an existing instance when it starts up, and then print the value of `existing` immediately before the call to `bs->InstallMultipleProtocolInterfaces()`.

**Comment by matatata** [comment]
I had the same error with my MacMini3,1. In src/interface/efi/efi_pxe.c I commented out the line:

```
 	if ( ( efirc = bs->InstallMultipleProtocolInterfaces (
 			&handle,
 			&efi_pxe_base_code_protocol_guid, &pxe->base,
			// -------------->>>> &efi_apple_net_boot_protocol_guid, &pxe->apple,
 			NULL ) ) != 0 ) {
 		rc = -EEFI ( efirc );
 		DBGC ( pxe, "PXE %s could not install base code protocol: %s\n",
```

and then I could even boot into  `chain --autofree http://boot.netboot.xyz/ipxe/netboot.xyz.efi` and then boot an AlpineLinux .
